components/TimeoutModal.tsx

```diff
@@ -5,76 +5,71 @@ import Modal from 'react-bootstrap/Modal'
 
 export interface TimeoutModalProps {
-  action: string
+  userArrivedFromUioMobile: boolean
   timedOut: boolean
 }
 
 export const TimeoutModal: React.FC<TimeoutModalProps> = (props) => {
   const { t } = useTranslation()
-  const { action, timedOut } = props
-  const TIMEOUT_MS = 30 * 60 * 1000
-  const TIMEOUT_DISPLAY_TIME_IN_MINUTES = 5
-  const TIMEOUT_WARNING_MS = TIMEOUT_MS - TIMEOUT_DISPLAY_TIME_IN_MINUTES * 60 * 1000
-  const [numberOfMinutes, setNumberOfMinutes] = useState(TIMEOUT_DISPLAY_TIME_IN_MINUTES)
-  const [showWarningModal, setShowWarningModal] = useState<boolean | null>(timedOut)
+  const { timedOut, userArrivedFromUioMobile } = props
 
-  useEffect(() => {
-    if (showWarningModal) {
-      const timer = setTimeout(() => {
-        setNumberOfMinutes(numberOfMinutes - 1)
-      }, 60 * 1000)
-      return () => clearTimeout(timer)
-    }
-  })
+  // handy converter for Minutes -> Milliseconds
+  const ONE_MINUTE_MS = 60 * 1000
+
+  // keep times in human readable minutes
+  const REDIRECT_TIMER = 30
+  const WARNING_DURATION = 5
+  const WARNING_TIMER = REDIRECT_TIMER - WARNING_DURATION
+
+  const [numberOfMinutes, setNumberOfMinutes] = useState(WARNING_DURATION)
+  const [showWarningModal, setShowWarningModal] = useState<boolean | null>(timedOut)
+  const [warned, setWarned] = useState<boolean | null>(timedOut)
 
   useEffect(() => {
     if (showWarningModal) {
       const timer = setTimeout(() => {
         setNumberOfMinutes(numberOfMinutes - 1)
-      }, 60 * 1000)
+      }, ONE_MINUTE_MS)
       return () => clearTimeout(timer)
     }
   })
 
-  function clear() {
-    if (timeOutTimerId) {
-      clearTimeout(timeOutTimerId)
-      timeOutTimerId = null
-    }
-    if (warningTimerId) {
-      clearTimeout(warningTimerId)
-      warningTimerId = null
-    }
-  }
-
-  function startOrUpdate() {
-    clear()
+  function startTimer() {
+    // Show the warning modal for a bit before navigating
     warningTimerId = setTimeout(() => {
       setShowWarningModal(true)
-      setNumberOfMinutes(TIMEOUT_DISPLAY_TIME_IN_MINUTES)
-    }, TIMEOUT_WARNING_MS)
-    timeOutTimerId = setTimeout(() => {
+      setWarned(true)
+      setNumberOfMinutes(WARNING_DURATION)
+    }, WARNING_TIMER * ONE_MINUTE_MS)
+    // And at the end, send back to EDD
+    setTimeout(() => {
       if (typeof window !== 'undefined') {
         const eddLocation = getUrl('edd-log-in')?.concat(encodeURIComponent(window.location.toString()))
         window.location.href = eddLocation || ''
       }
-    }, TIMEOUT_MS)
+    }, REDIRECT_TIMER * ONE_MINUTE_MS)
   }
 
   function closeWarningModal() {
     setShowWarningModal(false)
-    startOrUpdate()
+
+    if (warningTimerId) {
+      clearTimeout(warningTimerId)
+      warningTimerId = null
+    }
+  }
+
+  function redirectToUIHome() {
+    const uioHomeLink = userArrivedFromUioMobile ? getUrl('uio-home-url-mobile') : getUrl('uio-home-url-desktop')
+    window.location.href = uioHomeLink || ''
```

Comment:
Another change: instead of the button dismissing the modal, it lets the user head back to UI Home to keep their session active. This is only if they hit the button though - X'ing out the modal or clicking outside it lets them stay